components/sup/src/manager/service/pipe_hook_client.rs

```diff
@@ -0,0 +1,319 @@
+            for event in &events {
+                if event.readiness().contains(readiness) {
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    fn abs_pipe_name(&self) -> String { format!("\\\\.\\pipe\\{}", self.pipe_name) }
+
+    fn connect(&self) -> Result<(NamedPipe, Poll)> {
+        let mut opts = OpenOptions::new();
+        opts.read(true)
+            .write(true)
+            .custom_flags(winbase::FILE_FLAG_OVERLAPPED);
+        let file = opts.open(self.abs_pipe_name())?;
+
+        let pipe = unsafe { NamedPipe::from_raw_handle(file.into_raw_handle()) };
+        let poll = Poll::new()?;
+        poll.register(&pipe, Token(0), Ready::all(), PollOpt::edge())?;
+        Ok((pipe, poll))
+    }
+
+    fn init_server<T>(&self,
+                      service_group: &str,
+                      pkg: &Pkg,
+                      svc_encrypted_password: Option<T>)
+                      -> Result<()>
+        where T: ToString
+    {
+        let script_path = match henv::var(PIPE_SERVICE_ENVVAR) {
+            Ok(script) => PathBuf::from(script),
+            Err(_) => {
+                env::current_exe()?.parent()
+                                   .expect("unable to find sup binary parent dir")
+                                   .join("named_pipe_service.ps1")
+            }
+        };
+        let ps_cmd = format!("& '{}' -HookPath '{}' -PipeName {}",
+                             script_path.to_string_lossy(),
+                             self.hook_path.to_string_lossy(),
+                             self.pipe_name);
+        // Start instance of powershell to host named pipe server for this client
+
+        let args = vec!["-NonInteractive", "-Command", ps_cmd.as_str()];
+        let child = Child::spawn("pwsh.exe",
+                                 &args,
+                                 &pkg.env,
+                                 &pkg.svc_user,
+                                 svc_encrypted_password)?;
+        debug!("spawned powershell server for {} {} hook on pipe: {}",
+               service_group, self.hook_name, self.pipe_name);
+
+        let out = child.stdout;
+        let err = child.stderr;
+        let preamble_out_str = format!("{} hook[{}]:", service_group, self.hook_name);
+        let preamble_err_str = preamble_out_str.clone();
+        let out_path = self.stdout_log_file.clone();
+        let err_path = self.stderr_log_file.clone();
+        let handle = child.handle;
+
+        // The above powershell will run for the lifetime of this client which is the same
+        // duration as the service itself and will thus service each hook run. So we will
+        // stream its stdout/err in separate threads
+        thread::Builder::new().name(format!("{}-{}-pipe-server-out",
+                                            service_group, self.hook_name))
+                              .spawn(move || stream_output(out, &out_path, &preamble_out_str))
```

Comment:
OPTIONAL: there's no need to spawn threads in the case that `out` (or `err` in the subsequent code) is `None`. Moving that check here avoids consuming extra resources and simplifies the implementation of `stream_output`. It's not entirely clear to me whether we ever expect those streams to be `None`. If we don't, perhaps we should be generating an error rather than silently not logging.